### ISSUE TITLE
Повторное объявление при объявлении записи в пространствах имён

### ISSUE THREAD

**Issue by AIexandrKotov** [pascalabcnet/pascalabcnet#2025]
```
namespace Spacename;

type

  R = record // Повторно объявленный идентификатор R
  
  end;
  
end.
```
Не, ну это уже прямо совсем.

**Comment by miks1965** [comment]
Я предлагаю выключить этот поток сознания пока пространства имен не будут готовы и об этом как-то не будет официально объявлено.

**Comment by AIexandrKotov** [comment]
Тестирование уменьшит число багов, а если серьёзные недоделки замечать на ранних стадиях, то и работы, соответственно, делать меньше. Но я бы сказал, что пространствами имён пользоваться уже можно. Вот я и пользуюсь, попутно находя все эти мелкие баги. Что вас в этом не устраивает?

Лично мне без разницы на IDE-шную часть всего этого (тем более, что всё настраивается вручную, а я так-итак не любитель лёгких путей), которая как раз и является серьёзной недоделкой (но Intellisense уже работает). Часть компилятора же, как по мне, вполне юзабельна.

Чем больше issue по этой теме — тем быстрее пространства имён будут готовы. По-крайней мере, я на это надеюсь.

**Comment by ibond84** [comment]
> Я предлагаю выключить этот поток сознания пока пространства имен не будут готовы и об этом как-то не будет официально объявлено.

Нет. issues как раз необходимы. Разработка пространств имен после 2 лет застоя возобновлена. Товарищ, флудивший issue про неудаленные пробелы ушел. Можно теперь заняться чем-то более полезным и интересным

**Comment by miks1965** [comment]
Я согласен. Только раз активная разработка возобновилась, я буду настаивать чтобы пространства имен не привязывать к файлам, а сделать как в C#. Готов взяться за грамматику.

**Comment by ibond84** [comment]
Ну они и так не привязаны к файлам. Все сделано как в С#

**Comment by AIexandrKotov** [comment]
Скорее всего, miks1965 имел ввиду директиву {$includenamespace}. Если подключить файл, в котором используется класс, подключенный позже, то имя не найдёт. В C# с этим всё нормально. Но преимущество перед C# тут в том, что можно не создавать файл проекта. А чтобы убрать директиву {$includenamespace} так или иначе нужно всё делать исходя из .pabcproj.

**Comment by AIexandrKotov** [comment]
Как в C# в смысле в проектах? Ну, чтобы не прописывать в основном файле {$includenamespace}, сделать инфу в файле проекта по подключаемым пространствам имён и, соответственно, файлам.

**Comment by miks1965** [comment]
Я имею в виду  следующий синтаксис:
```
namespace A

end;

namespace B

end;

namespace System.Collections

end;

begin
end.
```

**Comment by AIexandrKotov** [comment]
У пространств имён не должно быть begin. Скорее уж так:
```
namespace A;

end.

namespace B;

end.

namespace System.Collections;

end.
```

**Comment by miks1965** [comment]
Точка - в конце файла. А это begin-end программы. Читайте внимательно

**Comment by ibond84** [comment]
>Я имею в виду следующий синтаксис:

Ну, это как раз то, от чего мы уходим. Пространства имен должны быть сквозными, чтобы не писать в одном файле сотни классов.

**Comment by miks1965** [comment]
Идею я понимаю. Но механизм подключения пространств имен через директивы компилятора - древний и восходит к #include-файлам в C, которые я критикую буквально на каждой лекции по C++. 

Кроме того, насколько я понимаю, в одном пространстве имен может быть сколько угодно классов.

Честно говоря, я не понимаю, почему мы уходим. В таких языках как C++, C#, Visual Basic пространства имен существуют ровно в том виде, в котором я предлагаю.

В Oxygene
https://en.wikipedia.org/wiki/Oxygene_(programming_language)#Code_examples
unit заменено на namespace, но сказано, что там нет пофайловой компиляции, а только компиляция всего проекта.

Как некий вариант предлагаю рассмотреть следующее: оставить один namespace в файле (как сейчас), но **запретить** компилировать namespace вне контекста проекта и **запретить** его подключение директивами компилятора. Тем самым будет достигнута аналогия: старый код Delphi будет компилироваться модулями как и в современных версиях Delphi, а явные namespaces будут возникать только в контексте проектов и компиляция будет не пофайловой, а попроектной - как и в C#.

**Comment by AIexandrKotov** [comment]
Я всегда любил PascalABC.NET за простоту, включая простоту многофайловости. Проекты в PascalABC.NET никогда не были обязательными и без них всегда всё запускалось. Даже библиотеки можно компилировать вне проекта. И C# я иногда компилирую именно через PascalABC.NET, ибо студия голые файлы без проекта не способна запускать. Это достоинство ЯП, следующее из его направленности на обучение. От этого нельзя отказываться.

> запретить компилировать namespace вне контекста проекта

Так уже есть этот запрет. Хотя слегка формальный. Просто пространство — скомпилировать нельзя. А вот обычный файл, к которому подключено пространство, даже без проекта — можно.
